Extensions/PrimitiveDrawing/shapepainterruntimeobject.ts

```diff
@@ -438,23 +450,294 @@ namespace gdjs {
+
+    getCenterY(): float {
+      if (!this._customCenter) {
+        return super.getCenterY();
+      }
+      let centerY =
+        this._customCenter[1] +
+        (this.getY() - this.getDrawableY()) / Math.abs(this._scaleY);
+      return centerY * Math.abs(this._scaleY);
+    }
+
+    /**
+     * Change the width of the object. This changes the scale on X axis of the object.
+     *
+     * @param width The new width of the object, in pixels.
+     */
+    setWidth(newWidth: float): void {
+      const unscaledWidth = this._renderer.getUnscaledWidth();
+      if (unscaledWidth !== 0) {
+        this.setScaleX(newWidth / unscaledWidth);
+      }
+    }
+
+    /**
+     * Change the height of the object. This changes the scale on Y axis of the object.
+     *
+     * @param height The new height of the object, in pixels.
+     */
+    setHeight(newHeight: float): void {
+      const unscaledHeight = this._renderer.getUnscaledHeight();
+      if (unscaledHeight !== 0) {
+        this.setScaleY(newHeight / unscaledHeight);
+      }
+    }
+
+    /**
+     * Change the scale on X and Y axis of the object.
+     *
+     * @param newScale The new scale (must be greater than 0).
+     */
+    setScale(newScale: float): void {
+      this.setScaleX(newScale);
+      this.setScaleY(newScale);
+    }
+
+    /**
+     * Change the scale on X axis of the object (changing its width).
+     *
+     * @param newScale The new scale (must be greater than 0).
+     */
+    setScaleX(newScale: float): void {
+      if (newScale < 0) {
+        newScale = 0;
+      }
+      if (newScale === Math.abs(this._scaleX)) {
+        return;
+      }
+      this._scaleX = newScale * (this._flippedX ? -1 : 1);
+      this._renderer.updateScaleX();
+      this.hitBoxesDirty = true;
+    }
+
+    /**
+     * Change the scale on Y axis of the object (changing its width).
+     *
+     * @param newScale The new scale (must be greater than 0).
+     */
+    setScaleY(newScale: float): void {
+      if (newScale < 0) {
+        newScale = 0;
+      }
+      if (newScale === Math.abs(this._scaleY)) {
+        return;
+      }
+      this._scaleY = newScale * (this._flippedY ? -1 : 1);
+      this._renderer.updateScaleY();
+      this.hitBoxesDirty = true;
+    }
+
+    flipX(enable: boolean): void {
+      if (enable !== this._flippedX) {
+        this._scaleX *= -1;
+        this._flippedX = enable;
+        this._renderer.updateScaleX();
+        this.hitBoxesDirty = true;
+      }
+    }
+
+    flipY(enable: boolean): void {
+      if (enable !== this._flippedY) {
+        this._scaleY *= -1;
+        this._flippedY = enable;
+        this._renderer.updateScaleY();
+        this.hitBoxesDirty = true;
+      }
+    }
+
+    isFlippedX(): boolean {
+      return this._flippedX;
+    }
+
+    isFlippedY(): boolean {
+      return this._flippedY;
+    }
+
+    /**
+     * Get the scale of the object (or the geometric mean of the X and Y scale in case they are different).
```

Comment:
The geometric mean gives the same value for the same surface:
Scale 4 is 4*4=16 times the base surface
Scale (1; 16) = 4 is 1*16=16 times the base surface

The square root is more costly but the value can actually be used whereas it's almost impossible to make sense out of a geometric mean.
For instance, it could be used to square a rectangular scaling: object.setScale(getScale()).

Reply:
If you're ok with this solution, do I harmonize it for the other objects in another PR?

Reply:
This would be a breaking change for other objects, so might need to think if this is worth it?